Update Arrow to 0.15.1 and fix Broadcast and GroupedMapUdf Tests for Spark-3.0.0

suhsteve opened the pull request:
This PR updates the Arrow library from 0.14.1 to 0.15.1 and also addresses the Broadcast and GroupedMapUdf Tests for Spark-3.0.0.  Currently supporting GroupedMapUdf in Spark-3.0.0 is blocked/unsupported so we skip these tests.


- Arrow 0.15.1 update #370 
  - Spark 3.0.0 uses 0.15.1 of Arrow and this version also introduces a new Arrow Spec [apache/arrow#4951](https://github.com/apache/arrow/pull/4951) and [apache/arrow@4be02c7](https://github.com/apache/arrow/commit/4be02c7a468055a9c1309fbeeca8985cc9810821)
  - As part of the new spec the end of stream message has also changed. 
 https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format for the new message format - 0xFFFFFFFF 0x00000000


This is a part of the effort to bring in CI for Spark 3.0: #348

imback82 commented:
Can you resolve conflicts? Thanks!

suhsteve commented:
> Can you resolve conflicts? Thanks!

Merged and resolved conflicts.  Re-enabled some tests.

imback82 commented:
@suhsteve is this ready for review?

suhsteve commented:
> @suhsteve is this ready for review?

yea, looks like tests are passing.

imback82 commented on azure-pipelines.yml:
```diff
@@ -16,6 +16,8 @@ variables:
   backwardCompatibleRelease: '0.9.0'
   TestsToFilterOut: "(FullyQualifiedName!=Microsoft.Spark.E2ETest.IpcTests.DataFrameTests.TestDataFrameGroupedMapUdf)&\
   (FullyQualifiedName!=Microsoft.Spark.E2ETest.IpcTests.DataFrameTests.TestDataFrameVectorUdf)&\
+  (FullyQualifiedName!=Microsoft.Spark.E2ETest.IpcTests.DataFrameTests.TestGroupedMapUdf)&\
```
This is a breaking change since these are not the new APIs introduced, no?

suhsteve replied:
Moved to spark 3.0.0 tests only.

imback82 commented on src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs:
```diff
@@ -763,33 +777,30 @@ internal class ArrowOrDataFrameGroupedMapCommandExecutor : ArrowOrDataFrameSqlCo
+                        writer = new ArrowStreamWriter(outputStream, result.Schema, true, ipcOptions);
                     }
 
-                    // TODO: Remove sync-over-async once WriteRecordBatch exists.
```
Do you know which line is removing this TODO?

imback82 commented on src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs:
```diff
@@ -763,33 +777,30 @@ internal class ArrowOrDataFrameGroupedMapCommandExecutor : ArrowOrDataFrameSqlCo
-            {
-                writer.Dispose();
-            }
+            writer.WriteEndAsync().GetAwaiter().GetResult();
```
`writer` can be null here?

imback82 commented on src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs:
```diff
@@ -30,12 +31,14 @@ internal abstract class SqlCommandExecutor
         /// Executes the commands on the input data read from input stream
         /// and writes results to the output stream.
         /// </summary>
+        /// <param name="version">Spark version.</param>
```
super nit: remove "." to be consistent.

imback82 commented on src/csharp/Microsoft.Spark.Worker.UnitTest/CommandExecutorTests.cs:
```diff
@@ -25,6 +26,19 @@ namespace Microsoft.Spark.Worker.UnitTest
+
+        public CommandExecutorTests()
+        {
+            // Spark 3.x and tests use ArrowStreamReader 0.15.1+
```
Shouldn't we also test against Spark 2?

suhsteve replied:
I guess the question is, how do we test against spark 2?  Spark 2 uses the legacy Arrow spec, however in these tests we're using Arrow 0.15.1, and therefore ArrowStreamReader is expecting the new Arrow spec.

imback82 replied:
set `WriteLegacyIpcFormat` to true when you test against Spark 2?

imback82 replied:
Arrow 0.15.1 is working against Spark 2, no?

suhsteve replied:
I just wasn't sure if `ArrowStreamReader` was capable of reading both legacy and non legacy specs.

suhsteve replied:
Ok, got it.   Thought we were only using ArrowStreamReader in the tests, but I guess we use it in the SqlCommandExecutor as well.  So it must be smart enough to read both.

imback82 commented on src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs:
```diff
@@ -763,33 +785,31 @@ internal class ArrowOrDataFrameGroupedMapCommandExecutor : ArrowOrDataFrameSqlCo
 
                     if (writer == null)
                     {
-                        writer = new ArrowStreamWriter(outputStream, result.Schema, leaveOpen: true);
```
ditto

imback82 commented:
@suhsteve Can you update the title/description with more details? This is more like upgrading Arrow version.

suhsteve commented on src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs:
```diff
@@ -387,7 +416,7 @@ internal class ArrowOrDataFrameSqlCommandExecutor : ArrowBasedCommandExecutor
                     Debug.Assert(resultSchema == null);
                     resultSchema = BuildSchema(results);
 
-                    writer = new ArrowStreamWriter(outputStream, resultSchema, leaveOpen: true);
```
readded.

suhsteve commented on src/csharp/Microsoft.Spark.Worker.UnitTest/CommandExecutorTests.cs:
```diff
@@ -25,9 +26,12 @@ namespace Microsoft.Spark.Worker.UnitTest
-        [Fact]
-        public void TestPicklingSqlCommandExecutorWithSingleCommand()
+        [Theory]
+        [InlineData(Versions.V2_4_2)]
```
Added comment and ignoring IpcOptions.

suhsteve commented on azure-pipelines.yml:
```diff
@@ -558,12 +558,16 @@ stages:
       env:
         SPARK_HOME: $(Build.BinariesDirectory)\spark-2.4.6-bin-hadoop2.7
 
+    # Spark 3.0.0 uses Arrow 0.15.1, which contains a new Arrow spec. This breaks backward
```
At the moment there are no published workers that's backward compatible with 3.0 (since the previous workers only use 0.14.1 and aren't aware of the new spec).  But I agree that this is a breaking change.

For backward compatibility, do we want to differentiate between different spark versions and test them against different spark Worker versions?  Or one Worker version where we say is backward compatible for all spark versions?

This can be addressed in a separate PR if needed.

imback82 replied:
> For backward compatibility, do we want to differentiate between different spark versions and test them against different spark Worker versions? Or one Worker version where we say is backward compatible for all spark versions?

I would say the latter.

suhsteve replied:
Then I think we will have to wait until the next official Worker release before we can remove these extra filters.

imback82 replied:
... since we have one worker binary to support all spark versions.

imback82 replied:
Well, you remove the backward compatibility test (breaking change) then added it back when the new one is released.

suhsteve replied:
Do you want me to remove the extra filters for 3.0 and in the unit tests add the skip attribute ?  Or just remove the spark 3.0.0 section in the backward compatibility tests.

suhsteve replied:
Removed 3.0 backward compatibility testing.

suhsteve commented:
> @suhsteve Can you update the title/description with more details? This is more like upgrading Arrow version.

Addressed the comments and updated title/description.

imback82 reviewed (dismissed):
LGTM, thanks @suhsteve!
